### PR TITLE
sdk/state: close channel

### PR DESCRIPTION
**Pull request by acharb**
this adds functionality to the state machine for closing a channel

closes https://github.com/stellar/experimental-payment-channels/issues/47

### CODE REVIEW

File: sdk/txbuild/declaration.go [hunk 1]
```diff
@@ -19,8 +19,9 @@ func Declaration(p DeclarationParams) (*txnbuild.Transaction, error) {
 			AccountID: p.InitiatorEscrow.Address(),
 			Sequence:  startSequenceOfIteration(p.StartSequence, p.IterationNumber) + 0, // Declaration is the first transaction in an iteration's transaction set.
 		},
-		BaseFee:           0,
-		Timebounds:        txnbuild.NewTimeout(300),
+		BaseFee: 0,
+		// TODO - Timebounds needs to be explicit
+		Timebounds:        txnbuild.NewInfiniteTimeout(),
 		MinSequenceNumber: &minSequenceNumber,
```
Comment (acharb) [inline sdk/txbuild/declaration.go:25]: this change was already made in `txbuild/close.go` , and needed to be copied here as well

File: sdk/state/state_test.go [hunk 2]
```diff
@@ -391,6 +414,177 @@ func Test(t *testing.T) {
...
+		require.True(t, fullySigned)
+		ci, di, err := sendingChannel.PaymentTxs(payment)
+		require.NoError(t, err)
+		ci, err = ci.AddSignatureDecorated(payment.CloseSignatures...)
+		require.NoError(t, err)
+		di, err = di.AddSignatureDecorated(payment.DeclarationSignatures...)
+		require.NoError(t, err)
+	}
+
+	// Coordinated Close
```
Comment (acharb) [inline sdk/state/state_test.go:529]: there are a lot of changes in this test file. But most of it is copying code/refactoring. The notable changes are in this section (testing a coordinated close).

Uncoordinated close is already being tested in `TestOpenUpdatesUncoordinatedClose`

File: sdk/state/state.go [hunk 3]
```diff
@@ -150,36 +174,6 @@ func (c *Channel) verifySigned(tx *txnbuild.Transaction, sigs []xdr.DecoratedSig
 	return false, nil
 }
 
-func (c *Channel) CloseStart(iterationNumber int) error {
```
Comment (acharb) [inline sdk/state/state.go:153]: this is removing functions that were originally added as part of the skeleton sdk. Now can be removed as the actual functions are being implemented

File: sdk/state/state_test.go [hunk 4]
```diff
@@ -17,29 +18,34 @@ import (
...
+		os.Exit(1)
 	}
+	networkPassphrase = networkDetails.NetworkPassphrase
 
-	// Setup initiator.
-	initiator := Participant{
+	os.Exit(m.Run())
+}
+
+func initAccounts(t *testing.T) (initiator Participant, responder Participant) {
```
Comment (nikhilsaraf) [inline sdk/state/state_test.go:47]: style: I'm generally discouraging of naming the return types in the signature (would rather use a comment), because that requires you to do a `=` assignment instead of a `:=` assignment which requires a code reader to check if that variable has been assigned anywhere above and whether it is used etc, essentially taking away from the "don't reuse variables" philosophy.

(another code style pattern that we don't have to adopt but thought I'd bring up in case we find it useful and choose to adopt)

File: sdk/state/state_test.go [hunk 5]
```diff
@@ -391,6 +414,177 @@ func Test(t *testing.T) {
...
+		_, err = client.SubmitFeeBumpTransaction(fbtx)
+		require.NoError(t, err)
+	}
+
+	// Perform a number of iterations, much like two participants may.
+	// Exchange signed C_i and D_i for each
+	t.Log("Subsequent agreements...")
+	rBalanceCheck := responder.Contribution
+	iBalanceCheck := initiator.Contribution
+	for i < 20 {
```
Comment (nikhilsaraf) [inline sdk/state/state_test.go:476]: why `20`?
can we put in a named variable?

File: sdk/state/state_test.go [hunk 6]
```diff
@@ -391,6 +414,177 @@ func Test(t *testing.T) {
...
+		i++
+		require.Equal(t, i, initiatorChannel.NextIterationNumber())
+		require.Equal(t, i, responderChannel.NextIterationNumber())
+		amount := randomPositiveInt64(t, 100_0000000)
```
Comment (nikhilsaraf) [inline sdk/state/state_test.go:480]: similarly here too, can we put the number in a named variable. helps with readability

File: sdk/state/state_test.go (hunk 5 above)
Comment (nikhilsaraf) [inline sdk/state/state_test.go:476]: @acharb this seems like non-deterministic testing, which doesn't make it very easy to replicate if it every fails.

Is there a way to make this into a deterministic testing pattern?
If yes, then can we add a TODO to convert this to a deterministic testing pattern?

tagging @leighmcculloch to get your thoughts as well.

File: sdk/state/state_test.go [hunk 7]
```diff
@@ -391,6 +414,177 @@ func Test(t *testing.T) {
 	assert.EqualValues(t, initiatorEscrowResponse.Balances[0].Balance, fmt.Sprintf("%.7f", float64(iBalanceCheck)/float64(1_000_0000)))
 }
 
+func TestOpenUpdatesCoordinatedClose(t *testing.T) {
```
Comment (nikhilsaraf) [inline sdk/state/state_test.go:417]: @acharb this test seems very long which makes it hard to follow, and much much harder to maintain.

not a blocker, but if we can break it up in a future PR soon then that would make it a little easier to follow/maintain. It may be a little more complicated to do and we may need to invest in developing a good test framework, but something we should discuss with @leighmcculloch once he's back.

File: sdk/state/state_test.go [hunk 8]
```diff
@@ -391,6 +414,177 @@ func Test(t *testing.T) {
...
+	require.NoError(t, err)
+	fbtx, err = fbtx.Sign(networkPassphrase, initiator.KP)
+	require.NoError(t, err)
+	_, err = client.SubmitFeeBumpTransaction(fbtx)
+	require.NoError(t, err)
+
+	t.Log("Initiator proposes a coordinated close")
+	cc, err := initiatorChannel.ProposeCoordinatedClose(0, 0)
+	require.NoError(t, err)
+	cc, err = responderChannel.ConfirmCoordinatedClose(cc)
```
Comment (nikhilsaraf) [inline sdk/state/state_test.go:551]: 💯  on this style of testing remote/local channels without requiring networking logic. 🎉

File: sdk/state/state_test.go [hunk 9]
```diff
@@ -391,6 +414,177 @@ func Test(t *testing.T) {
...
+		BaseFee:    txnbuild.MinBaseFee,
+	})
+	require.NoError(t, err)
+	fbtx, err = fbtx.Sign(networkPassphrase, initiator.KP)
+	require.NoError(t, err)
+	_, err = client.SubmitFeeBumpTransaction(fbtx)
+	if err != nil {
+		hErr := horizonclient.GetError(err)
+		t.Log("Submitting Close:", txCoordinated.SourceAccount().Sequence, "Error:", err)
+		t.Log(hErr.ResultString())
```
Comment (nikhilsaraf) [inline sdk/state/state_test.go:572]: do we need to return if this errors out, or is it meaningful to continue through the end of this test?

File: sdk/state/state_test.go [hunk 10]
```diff
@@ -391,6 +414,177 @@ func Test(t *testing.T) {
...
+	}
+	t.Log("Coordinated close successful")
+
+	// check final escrow fund amounts are correct
+	accountRequest := horizonclient.AccountRequest{AccountID: responder.Escrow.Address()}
+	responderEscrowResponse, err := client.AccountDetail(accountRequest)
```
Comment (nikhilsaraf) [inline sdk/state/state_test.go:578]: just noticed that we are making a call to horizon. It is best if tests can avoid making calls to an external network, because we don't know if that network will be available during the running of this test.

do we need to make this call or can we check if the value is equal to what we should have initialized it to in the beginning of the test, without making the call to horizon?

**Review by nikhilsaraf — APPROVED** [review]
Added a bunch of comments.

None of them are blocking so marked as approved. But let's discuss in person, may be quickest.

File: sdk/state/close.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+	observationPeriodTime      time.Duration
+	observationPeriodLedgerGap int64
+	closeSignatures            []xdr.DecoratedSignature
+}
+
+func (cc CoordinatedClose) CloseSignatures() []xdr.DecoratedSignature {
+	return cc.closeSignatures
+}
+
+func (c *Channel) CloseTxs() (*txnbuild.Transaction, *txnbuild.Transaction, error) {
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/close.go]: good idea. I haven't seen comment vars used in the sdf go code before, but I added named return variables. I'd be happy to discuss further though
https://github.com/stellar/experimental-payment-channels/pull/83/commits/a08e853171efce60fac09a2fa59077578749756b

File: sdk/state/close.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+	txClose, err := txbuild.Close(txbuild.CloseParams{
+		ObservationPeriodTime:      c.observationPeriodTime,
+		ObservationPeriodLedgerGap: c.observationPeriodLedgerGap,
+		InitiatorSigner:            c.initiatorSigner(),
+		ResponderSigner:            c.responderSigner(),
+		InitiatorEscrow:            c.initiatorEscrowAccount().Address,
+		ResponderEscrow:            c.responderEscrowAccount().Address,
+		StartSequence:              c.startingSequence,
+		IterationNumber:            c.latestCloseAgreement.IterationNumber,
+		AmountToInitiator:          c.initiatorClaimAmount(),
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/close.go:48]: ya  you're right there will most likely need to be locking. I think it would be best to add that when we start thinking about networking, since that's when it will most be needed.

I added a ticket to track that thought: https://github.com/stellar/experimental-payment-channels/issues/84

File: sdk/state/close.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+		AmountToResponder:          c.responderClaimAmount(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return txClose, nil
+}
+
+// ProposeClose proposes parameters for a close transaction to be submitted earlier.
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/close.go]: https://github.com/stellar/experimental-payment-channels/pull/83/commits/e48d2ad592400d201bf2a94a78ff2db5af4cf5fd

File: sdk/state/close.go [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+		ResponderEscrow:            c.responderEscrowAccount().Address,
+		StartSequence:              c.startingSequence,
+		IterationNumber:            c.latestCloseAgreement.IterationNumber,
+		AmountToInitiator:          c.initiatorClaimAmount(),
+		AmountToResponder:          c.responderClaimAmount(),
+	})
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+	txCoordinatedClose, err = txCoordinatedClose.Sign(c.networkPassphrase, c.localSigner)
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/close.go:95]: good point. Personally I think changing names for the same struct adds some reading friction, eg. `txCoordinatedClose` vs `txCoordiantedCloseSigned`. I'd be happy to discuss though. We've been using the same name vars so far

File: sdk/state/close.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	return CoordinatedClose{
+		observationPeriodTime:      observationPeriodTime,
+		observationPeriodLedgerGap: observationPeriodLedgerGap,
+		closeSignatures:            txCoordinatedClose.Signatures(),
+	}, nil
+}
+
+func (c *Channel) ConfirmCoordinatedClose(cc CoordinatedClose) (CoordinatedClose, error) {
+	txCoordinatedClose, err := txbuild.Close(txbuild.CloseParams{
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/close.go]: I think that will help with readability 👍  added
https://github.com/stellar/experimental-payment-channels/pull/83/commits/a5df1ce03c8f422f0320f31ed6eafc3b69cd33ac

File: sdk/state/close.go [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+		return CoordinatedClose{}, err
+	}
+
+	// If remote has not signed coordinated close, error as is invalid.
+	signed, err := c.verifySigned(txCoordinatedClose, cc.closeSignatures, c.remoteSigner)
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+	if !signed {
+		return CoordinatedClose{}, fmt.Errorf("verifying coordinated close signed by remote: not signed by remote")
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/close.go]: this is currently the pattern we've adopted for errors so far (eg. if you look in `update.go`), but I can see it being confusing 🤔  basically we're trying to mimic wrapped errors. Probably near the finish line we can look at verification errors more holistically to come up with a better/clearer pattern. I changed this one to read clearer though
https://github.com/stellar/experimental-payment-channels/pull/83/commits/37cfbc4afdde35326c55276de00cbeeb10996d7d

File: sdk/state/close.go [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	// If local has not signed, sign.
+	signed, err = c.verifySigned(txCoordinatedClose, cc.closeSignatures, c.localSigner)
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+	if !signed {
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/close.go:93]: So the general steps are:

```
// 1. Initiator or Responder submits latest declaration tx
// (in steps 2-4 Initiator and Responder are interchangeable, as long as they alternate)
// 2. Initiator calls ProposeCoordinatedClose
// 3. Responder calls ConfirmCoordinatedClose
// 4. Initiator calls ConfirmCoordinatedClose
```
So in step 4 initiator (local) has already signed. They sign it when they propose it.

This shouldn't be a bug though. If you're the initiator, when you propose the new `txClose` you sign it and send to the responder. If the responder submits it, great that's what you wanted. If responder doesn't, then you can still just wait for `O` to pass and submit the original `txClose` that you already have fully signed.

the coordinated close has the same data as the latest `txClose` just with the shortened `O` period

happy to discuss further

File: sdk/state/state.go [hunk 18]
```diff
@@ -133,6 +143,20 @@ func (c *Channel) responderSigner() *keypair.FromAddress {
 	}
 }
 
+func (c *Channel) initiatorClaimAmount() int64 {
+	if c.latestCloseAgreement.Balance.Amount < 0 {
+		return c.latestCloseAgreement.Balance.Amount * -1
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/state.go:148]: correct if the channel balance is positive it means owed to responder. If it's negative it means owed to initiator. We added a comment for that here: https://github.com/stellar/experimental-payment-channels/blob/a946b6b9b45735f7b4270542ec0f699c06a5c699/sdk/state/state.go#L83

File: sdk/state/state_test.go (hunk 4 above)
Comment (acharb) [reply sdk/state/state_test.go:47]: I see your point thx for pointing out 🤔 . I'd be happy to discuss but the comment solution feels weird to me. I might be just not used to it though. I think when Leigh gets back who has more golang experience than I he can give his input here.

I think I'll leave as is for now if alright 😬

File: sdk/state/state_test.go (hunk 5 above)
Comment (acharb) [reply sdk/state/state_test.go]: 20 just felt like a good number of iterations to make sure the back and forth updates is working correctly. I made it a named var though:
https://github.com/stellar/experimental-payment-channels/pull/83/commits/e08f796e1f3771b5b549865c01737ba702d9cba7

File: sdk/state/state_test.go (hunk 6 above)
Comment (acharb) [reply sdk/state/state_test.go:484]: added a comment, I think that should be clearer 😄 
https://github.com/stellar/experimental-payment-channels/pull/83/commits/d48f3be031f4512fbed2f78979b6259366b4a724

File: sdk/state/state_test.go (hunk 5 above)
Comment (acharb) [reply sdk/state/state_test.go]: i don't feel strongly here. The only non-deterministic parts of these integration tests are the payment amounts and that we're calling a local standalone network. I think for this use case and stage of the project that makes sense, but I'd be happy to discuss further. 

We have a ticket [here](https://github.com/stellar/experimental-payment-channels/issues/70) for adding unit testing which won't rely on a network and can have non-random payment amounts to be deterministic

File: sdk/state/state_test.go (hunk 7 above)
Comment (acharb) [reply sdk/state/state_test.go:419]: I agree. This state_test.go can be cleaned up. I couldn't figure out a good way to write this second test w/o duplicating a lot of the code in the original test. 

I think once the majority of the sdf/state code is written this integration test can be broken up / looked at holistically and made more modular so there's less dup code. I didn't think now is the right time to do that though since stuff is changing frequently because early in the project

File: sdk/state/state_test.go (hunk 9 above)
Comment (acharb) [reply sdk/state/state_test.go:576]: oops good catch, we should end it
https://github.com/stellar/experimental-payment-channels/pull/83/commits/5fa3ee4c71d32837169639010cf138ec38b71988

File: sdk/state/state_test.go (hunk 10 above)
Comment (acharb) [reply sdk/state/state_test.go:583]: for this test file we're doing an integration test with a standalone network that includes CAP-21. This makes sure the transactions are formatted correct with correct ops/values and can be submitted successfully, among other things like making sure sequence numbers are correct.

We have a [ticket](https://github.com/stellar/experimental-payment-channels/issues/70) for adding unit tests, which won't depend on a network call

**Comment by acharb** [comment]
@nikhilsaraf got over eager and just responded to all your comments now 😄 . But happy to discuss still, lmk if you want to. Going to merge this in now though

File: sdk/state/close.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (nikhilsaraf) [reply sdk/state/close.go:62]: haha I like having those two names being separated like in your example :)
I'm ok with the way it is and don't want to push too hard if we're reused across the codebase.

File: sdk/state/close.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (nikhilsaraf) [reply sdk/state/close.go:93]: that makes sense since it's ok for responder to submit earlier since this is a coordinated close.

wondering if it's already signed in step 2, then why do we need to check if it is !signed in step 4?
(basically I'm wondering if we are checking for a state that isn't possible. and if that state occurs then it's a programmer error). can discuss on call.

File: sdk/state/close.go [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+	"github.com/stellar/experimental-payment-channels/sdk/txbuild"
+	"github.com/stellar/go/txnbuild"
+	"github.com/stellar/go/xdr"
+)
+
+// The steps for a channel coordinated close are as followed:
+// 1. Initiator or Responder submits latest declaration tx
+// (in steps 2-4 Initiator and Responder are interchangeable, as long as they alternate)
+// 2. Initiator calls ProposeCoordinatedClose
```
Comment (leighmcculloch) [inline sdk/state/close.go:15]: Why does the initiator have to call this step? Either participant should be able to take the first step.

File: sdk/state/close.go [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+// 3. Responder calls ConfirmCoordinatedClose
+// 4. Initiator calls ConfirmCoordinatedClose
+
+type CoordinatedClose struct {
+	observationPeriodTime      time.Duration
+	observationPeriodLedgerGap int64
```
Comment (leighmcculloch) [inline sdk/state/close.go:21]: These periods of time should be zero for a coordinated close. Why are they required?

File: sdk/state/close.go [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+	closeSignatures            []xdr.DecoratedSignature
```
Comment (leighmcculloch) [inline sdk/state/close.go:22]: The other objects we have for Payment, Open, keep all their fields exported so that they're inspectable, and serializable. We should do the same here too I think, at least for consistency, but also so we have the same benefits and pattern for use.

File: sdk/state/close.go [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+func (c *Channel) ConfirmCoordinatedClose(cc CoordinatedClose) (CoordinatedClose, error) {
+	txCoordinatedClose, err := c.makeCloseTx(cc.observationPeriodTime, cc.observationPeriodLedgerGap)
+	if err != nil {
+		return CoordinatedClose{}, err
+	}
+
+	// If remote has not signed coordinated close, error as is invalid.
+	signed, err := c.verifySigned(txCoordinatedClose, cc.closeSignatures, c.remoteSigner)
+	if err != nil {
+		return CoordinatedClose{}, err
```
Comment (leighmcculloch) [inline sdk/state/close.go:82]: We should wrap `err` here with `fmt.Errorf` like we do in other places so that if an error occurs we know what was happening when it occurred.

E.g.: https://github.com/stellar/experimental-payment-channels/blob/c30a604c216a5d8015e20309f3aad4d20ef67fb6/sdk/state/open.go#L120

File: sdk/state/close.go [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+		return CoordinatedClose{}, nil
+	}
+	return CoordinatedClose{
+		observationPeriodTime:      observationPeriodTime,
+		observationPeriodLedgerGap: observationPeriodLedgerGap,
+		closeSignatures:            txCoordinatedClose.Signatures(),
+	}, nil
+}
+
+func (c *Channel) ConfirmCoordinatedClose(cc CoordinatedClose) (CoordinatedClose, error) {
```
Comment (leighmcculloch) [inline sdk/state/close.go:73]: Can we return a fullySigned bool like we do in other propose/confirm function pairs?

E.g.:
https://github.com/stellar/experimental-payment-channels/blob/c30a604c216a5d8015e20309f3aad4d20ef67fb6/sdk/state/open.go#L109

File: sdk/state/state.go [hunk 24]
```diff
@@ -43,6 +43,8 @@ type Channel struct {
 
 	latestCloseAgreement     CloseAgreement
 	latestUnconfirmedPayment Payment
+
+	coordinatedClose CoordinatedClose
```
Comment (leighmcculloch) [inline sdk/state/state.go:47]: Coordinating a close is really just modifying the previous close agreement. I think it would be helpful because of that if we store the signatures for it in a CloseAgreement in the latestCloseAgreement. This way the last close is still there in the same place, it can just be executed early.

I think this would require adding the observation periods to the close agreement, and so merging the CoordinatedClose and CloseAgreement types.

Wdyt?

File: sdk/state/state.go [hunk 25]
```diff
@@ -133,6 +143,20 @@ func (c *Channel) responderSigner() *keypair.FromAddress {
 	}
 }
 
+func (c *Channel) initiatorClaimAmount() int64 {
```
Comment (leighmcculloch) [inline sdk/state/state.go:146]: We should name these functions `initiatorBalanceAmount` using the term Balance instead of Claim, since we don't use the term claim anywhere else.

File: sdk/state/state_test.go (hunk 7 above)
Comment (leighmcculloch) [reply sdk/state/state_test.go:419]: > w/o duplicating a lot of the code

We don't need to keep our tests DRY. Let's keep tests isolated from each other. It makes it easier to delete them when we don't need them, or for them to evolve independently of each other.

Also, this is likely to get better over time because the more functionality we add to the SDK, the tests will get naturally smaller as functionality moves out of these integration tests into the SDK.

**Review by leighmcculloch — COMMENTED** [review]
Looks great. I left a few comments below.

File: sdk/state/close.go (hunk 19 above)
Comment (acharb) [reply sdk/state/close.go:15]: ya  it can be either, above step 2 in parentheses it says that. I'll move it down a line so that sticks out more

File: sdk/state/close.go (hunk 20 above)
Comment (acharb) [reply sdk/state/close.go:21]: will a coordinated close always have a zero observation time? I wasn't sure if a shorter O (not necessarily zero) should be allowed

looking at the sep again though that it what is says. I can change

File: sdk/state/state.go (hunk 24 above)
Comment (acharb) [reply sdk/state/state.go:47]: But there would be two kinds of closeSignatures in that case right? eg:
```
type CloseAgreement struct {
	IterationNumber       int64
	Balance               Amount
	CloseSignatures       []xdr.DecoratedSignature
	DeclarationSignatures []xdr.DecoratedSignature

        // new:
	CoordinatedCloseSignature []xdr.DecoratedSignature
}
```
That felt awkward to me when I tried that, so created another type for separation of concerns.  🤔

File: sdk/state/close.go (hunk 21 above)
Comment (acharb) [reply sdk/state/close.go:22]: https://github.com/stellar/experimental-payment-channels/pull/95/commits/7b8efe8cb40017fd90aca14577e7effca343155b

File: sdk/state/close.go (hunk 22 above)
Comment (acharb) [reply sdk/state/close.go:82]: https://github.com/stellar/experimental-payment-channels/pull/95/commits/ac3ee05b8c4e42b0d3eec6522f5d4339dd26c6d2

File: sdk/state/state.go (hunk 25 above)
Comment (acharb) [reply sdk/state/state.go:146]: https://github.com/stellar/experimental-payment-channels/pull/95/commits/02f9d38f37cd5e88b3d0c739b0bb9debb6753faf

File: sdk/state/close.go (hunk 23 above)
Comment (acharb) [reply sdk/state/close.go:73]: https://github.com/stellar/experimental-payment-channels/pull/95/commits/66759128404aae82555bbdc864396e7448c13b65

File: sdk/state/close.go (hunk 20 above)
Comment (acharb) [reply sdk/state/close.go:21]: https://github.com/stellar/experimental-payment-channels/pull/95/commits/e00b88157d4faf1510792a74ce340487cdec34fc

File: sdk/state/close.go (hunk 19 above)
Comment (acharb) [reply sdk/state/close.go:15]: https://github.com/stellar/experimental-payment-channels/pull/95/commits/b5083d0128fb1a48da68137f412f009fe198d3ee

File: sdk/state/state.go (hunk 24 above)
Comment (leighmcculloch) [reply sdk/state/state.go:47]: Hmm, good point. We'd also be mutating an object that conceptually was locked in previously.

I think the thing I'm stuck on is that the channel has only two states we should be tracking:
- The most recent fully signed agreement.
- The agreement we're considering right now, and that agreement is either a new agreement (a new payment) of the adjustment of some attribute of the most recent agreement.

If the `lastUnconfirmedAgreement` was a `CloseAgreement` rather than a `Payment`, and if the `CloseAgreement` stored the observation periods then that would probably work. 

Storing them separately works too, but I worry we've now go to check in a few places if a coordinate close or a payment is in progress, because if either is in progress we shouldn't let the other begin, that could be dangerous.

Comment (acharb) [reply sdk/state/state.go:47]: > Storing them separately works too, but I worry we've now go to check in a few places if a coordinate close or a payment is in progress, because if either is in progress we shouldn't let the other begin, that could be dangerous.

that's true. I'm changing the close code to work without a `CoordinatedClose` type to see how it looks. I'll have a PR out
